apps/passport-server/src/services/generic-issuance/AutoIssuanceProvider.ts

```diff
@@ -0,0 +1,159 @@
+          realTickets
+        ))
+      );
+    }
+
+    return newManualTickets;
+  }
+
+  public async maybeIssueForUser(
+    userEmail: string,
+    existingManualTickets: ManualTicket[],
+    realTickets: PretixAtom[]
+  ): Promise<ManualTicket[]> {
+    const userRealTickets = realTickets.filter((t) => t.email === userEmail);
+    const userManualTickets = existingManualTickets.filter(
+      (t) => t.attendeeEmail === userEmail
+    );
+    const newManualTickets: ManualTicket[] = [];
+
+    for (const autoIssuance of this.autoIssuanceConfig) {
+      const permissioningRealTicket = userRealTickets.find((t) =>
+        ticketMatchesCriteria(t, autoIssuance.memberCriteria)
+      );
+      const permissioningManualTicket = userManualTickets.find((t) =>
+        ticketMatchesCriteria(t, autoIssuance.memberCriteria)
+      );
+
+      const matchesIssuanceMembershipCriteria: boolean =
+        !!permissioningRealTicket || !!permissioningManualTicket;
+
+      if (!matchesIssuanceMembershipCriteria) {
+        continue;
+      }
+
+      if (
+        !canIssueInThisEpoch(autoIssuance, existingManualTickets, userEmail)
+      ) {
+        continue;
+      }
+
+      const now = new Date();
+      for (let i = 0; i < autoIssuance.quantity; i++) {
+        const newManualTicket: ManualTicket = {
+          attendeeEmail: userEmail,
+          attendeeName:
+            permissioningRealTicket?.name ??
+            permissioningManualTicket?.attendeeName ??
+            "no name",
+          eventId: autoIssuance.eventId,
+          productId: autoIssuance.productId,
+          id: randomUUID(),
+          timeCreated: now.toISOString()
+        };
+
+        newManualTickets.push(newManualTicket);
+      }
+    }
+
+    logger(
+      `[AUTO_ISSUANCE] auto-issuing ${newManualTickets.length} tickets for user ${userEmail}`
+    );
+
+    return newManualTickets;
+  }
+}
+
+export function anyTicketMatchesCriteria(
+  ts: Array<PretixAtom | ManualTicket>,
+  criterias: MemberCriteria[]
+): boolean {
+  return !!ts.find((t) => ticketMatchesCriteria(t, criterias));
+}
+
+export function ticketMatchesCriteria(
+  t: PretixAtom | ManualTicket,
+  criterias: MemberCriteria[]
+): boolean {
+  return !!criterias.find((c) => {
+    if (t.eventId !== c.eventId) {
+      return false;
+    }
+    if (c.productId && c.productId !== t.productId) {
+      return false;
+    }
+    return true;
+  });
+}
+
+function canIssueInThisEpoch(
+  autoIssuance: AutoIssuanceOptions,
+  manualTickets: ManualTicket[],
+  email: string
+): boolean {
+  const start = new Date(autoIssuance.schedule.startDate).getTime();
+  const end = new Date(
+    autoIssuance.schedule.endDate ?? new Date(3000, 1, 1)
+  ).getTime();
+  const now = new Date().getTime();
+
+  if (now < start || now >= end) {
+    return false;
+  }
+
+  const ticketsForUserInEpoch = manualTickets.filter((t) => {
+    if (t.attendeeEmail !== email) {
+      return false;
+    }
+
+    if (!t.timeCreated) {
```

Comment:
Does this mean "manual ticket that was created from JSON configuration in the Pipeline definition"? If it does, a comment for this would be good.